docs/site/Life-cycle.md

```diff
@@ -0,0 +1,217 @@
+
+It's often desirable for various types of artifacts to participate in the life
+cycles and perform related processing upon `start` and `stop`. Good examples of
+such artifacts are:
+
+- Servers
+
+  - start: Starts the HTTP server listening for connections.
+  - stop: Stops the server from accepting new connections.
+
+- Components
+
+  - A component can register life cycle observers
+
+- DataSources
+
+  - connect: Connect to the underlying database or service
+  - disconnect: Disconnect from the underlying database or service
+
+- Custom scripts
+  - start: Custom logic to be invoked when the application starts
+  - stop: Custom logic to be invoked when the application stops
+
+## The `LifeCycleObserver` interface
+
+To react on life cycle events, a life cycle observer implements the
+`LifeCycleObserver` interface.
+
+```ts
+import {ValueOrPromise} from '@loopback/context';
+
+/**
+ * Observers to handle life cycle start/stop events
+ */
+export interface LifeCycleObserver {
+  start?(): ValueOrPromise<void>;
+  stop?(): ValueOrPromise<void>;
+}
+```
+
+Please note all methods are optional so that an observer can opt in certain
+events. Each main events such as `start` and `stop` are further divided into
+three sub-phases to allow the multiple-step processing.
+
+## Register a life cycle observer
+
+A life cycle observer can be registered by calling `lifeCycleObserver()` of the
+application. It binds the observer to the application context with a special
+tag - `CoreTags.LIFE_CYCLE_OBSERVER`.
+
+```ts
+app.lifeCycleObserver(MyObserver);
+```
+
+Please note that `app.server()` automatically registers servers as life cycle
+observers.
+
+Life cycle observers can be registered via a component too:
+
+```ts
+export class MyComponentWithObservers {
+  lifeCycleObservers = [XObserver, YObserver];
+}
+```
+
+## Discover life cycle observers
+
+The `Application` finds all bindings tagged with `CoreTags.LIFE_CYCLE_OBSERVER`
+within the context chain and resolve them as observers to be notified.
+
+## Notify life cycle observers of start/stop related events by order
+
+There may be dependencies between life cycle observers and their order of
+processing for `start` and `stop` need to be coordinated. For example, we
+usually start a server to listen on incoming requests only after other parts of
+the application are ready to handle requests. The stop sequence is typically
+processed in the reverse order. To support such cases, we introduce
+two-dimension steps to control the order of life cycle actions.
+
+### Observer groups
+
+First of all, we allow each of the life cycle observers to be tagged with a
+group. For example:
+
+- datasource (connect/disconnect)
+
+  - mongodb
+  - mysql
+
+- server
+  - rest
+  - gRPC
+
+We can then configure the application to trigger observers group by group as
+configured by an array of groups in order such as `['datasource', 'server']`.
+Observers within the same group can be notified in parallel.
```

Comment:
> Can be notified in parallel

I find this part ambiguous. Are you trying to say that observers are notified in series by default, but there is an option to enable parallel processing instead? How can readers of this paragraph find how to set that option?

Or did you perhaps want to say that _observers within the same group **are** notified in parallel_?